### PR TITLE
Avoid instantiating method not found exceptions when trying out properties

### LINKED ISSUE #20303: UserPropertyMapper generated exceptions on mapping
### Description

The UserPropertyMapper allows users to map user information into a token. The implementation cause a performance slowdown as it constantly creates exceptions when a property is not found when trying different formats. 

At the same time it doesn't provide error messages to users when they specify an illegal value, and misconfigurations are hard to spot. 

Therefore this area needs improvements. 

### Discussion

_No response_

### Motivation

The problems have been spotted when doing a performance review for KC22. It would be great to have this out-of-the-way to for future performance evaluations.

### Details

_No response_

### PR DESCRIPTION
**Pull request by ahus1**
This is a performance optimization and also improves usability in the UI as it validates properties.

Closes #20303

### CODE REVIEW

**Comment by ahus1** [comment]
@mhajas - this is now ready for review. It turned out bigger than anticipated with the additional tests and the migration, so I suppose it was a good call to move it a separate issue.

**Comment by ahus1** [comment]
@mhajas: I just noticed that `UserPropertyAttributeStatementMapper` for SAML is affected by the change in `ProtocolMapperUtils` as well, will update it accordingly. Still like to get some early preview if you have the time.

**Comment by ahus1** [comment]
@mhajas - I now pushed a similar change for the SAML part

File: model/legacy-private/src/main/java/org/keycloak/migration/migrators/MigrateTo22_0_0.java
Comment (mhajas) [inline model/legacy-private/src/main/java/org/keycloak/migration/migrators/MigrateTo22_0_0.java:37]: Should we do the Liquibase migration for this instead? The advantage is that we don't need to load all realms and all client scopes, but we can update only those we are interested in. See for example this:
https://github.com/keycloak/keycloak/blob/main/model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate21_0_2_TermsAndConditionsRequiredAction.java

File: testsuite/model/src/test/java/org/keycloak/testsuite/model/clientscope/ClientScopeModelTest.java
Comment (mhajas) [inline testsuite/model/src/test/java/org/keycloak/testsuite/model/clientscope/ClientScopeModelTest.java:113]: I like the idea of having the tests like this, however, if we change the migration to Liquibase one, this would not work.

If we decide to have the migration the current way, we should introduce a separate migration test class so it is in one place.

**Review by mhajas — COMMENTED** [review]
Thank you for PR I added some comments to the code.

File: model/legacy-private/src/main/java/org/keycloak/migration/migrators/MigrateTo22_0_0.java
Comment (ahus1) [reply model/legacy-private/src/main/java/org/keycloak/migration/migrators/MigrateTo22_0_0.java]: The Liquibase migration turned out to be complicated as it would manipulate a CLOB field which needs different in the various databases. Liquibase's Java API also didn't allow for calculated values. 

The updated version of the code now handles old and new values, and old values print a warning.

File: testsuite/model/src/test/java/org/keycloak/testsuite/model/clientscope/ClientScopeModelTest.java
Comment (ahus1) [reply testsuite/model/src/test/java/org/keycloak/testsuite/model/clientscope/ClientScopeModelTest.java]: Resolved, as there is no migration any more.

**Comment by ahus1** [comment]
@mhajas - this is now ready for another review: There is no longer migration necessary. This follows your initial idea of supporting both versions. I rejected it early as I thought it would have a runtime overhead to compare String variants at runtime. Now I found a way to avoid a runtime overhead. 

Please approve when you consider this ready to be merged. Thanks!

**Comment by ahus1** [comment]
@mposolda - This PR originated from a KC22 performance review. The change has already been discussed with @mhajas, and I asked him for a review.

I'm adding you as a reviewer as it touches Keycloak Core - please feel free to unassign your self or ask someone else from the core team for a review. Thanks!

**Review by mhajas — DISMISSED** [review]
Thanks @ahus1, LGTM

**Comment by pedroigor** [comment]
@ahus1 FTR, we recently merged https://github.com/keycloak/keycloak/pull/20614.

We are discussing going even further and avoiding any reflection at runtime when mapping claims to tokens.

**Comment by ahus1** [comment]
@pedroigor - I had a look at the other PR #20614, and I see that it is using the attribute mapper more often, still it doesn't migrate old usages or the property mapper or removes it completely. Therefore I think this PR is still valid, as it also improves the usability of this mapper. 

I'm still waiting for an approval of the @keycloak/core team for this. I've now rebased it as there was a conflict on the release notes.

**Review by mhajas — APPROVED** [review]
Re-approving

**Comment by pedroigor** [comment]
@ahus1 I'm not sure if it is valid. We should work to removing reflection not to make it perform better.

It is a green-field change (touching a sensitive area) and the idea is to evaluate first before introducing more steps during upgrades/migration.

**Comment by ahus1** [comment]
Hi @pedroigor - thanks for letting me know. I'm closing this issue and PR and leave it to the core team to follow up on the bits as it sees fit.

**Review by mposolda — COMMENTED** [review]
This change is probably an improvement, however I also wonder if we can improve more and remove reflection entirely? Something like:

```
private static final Map<String, Function<UserModel, String>> USER_MODEL_METHODS = new HashMap()<>(); // or ConcurrentHashMap? May depend on whether lazy init or not...

{
  // initialize somehow (not sure if lazy init or eager startup. Eager startup here for illustration...
  USER_MODEL_METHODS.put(UserModel.FIRST_NAME, UserModel::getFirstName);
  USER_MODEL_METHODS.put(UserModel.LAST_NAME, UserModel::getLastName);
 ...
}

```

BTV. One thing to note that there is this recent change https://github.com/keycloak/keycloak/commit/c22972af, which makes the use of `UserPropertyMapper` removed by default and replaced by `UserAttributeMapper`. But still it doesn't affect the "custom" properties added by users.

**Comment by ahus1** [comment]
@mposolda - this PR is now closed, and I don't have plans to re-open it as I'm no longer focusing on the KC22 vs. RHSSO 7.6 performance. Feel free to re-use the bits and pieces as necessary in other PRs created by the core team.

**Pull request closed without merging**
